lib/firestore/delete.js

```diff
@@ -280,18 +273,120 @@ FirestoreDelete.progressBar = new ProgressBar("Deleted :current docs (:rate docs
  */
 FirestoreDelete.prototype._recursiveBatchDelete = function() {
   var self = this;
-  return this._getDescendantBatch(this.allDescendants, this.batchSize).then(function(docs) {
-    if (docs.length <= 0) {
-      return Promise.resolve();
+
+  // Tunable deletion parameters
+  var readBatchSize = 7500;
+  var deleteBatchSize = 250;
+  var maxPendingDeletes = 15;
+  var maxQueueSize = deleteBatchSize * maxPendingDeletes * 2;
+
+  // All temporary variables for the deletion queue.
+  var queue = [];
+  var numPendingDeletes = 0;
+  var pagesRemaining = true;
+  var pageIncoming = false;
+  var lastDocName;
+
+  var failures = [];
+  var retried = {};
+
+  var queueLoop = function() {
+    if (queue.length == 0 && numPendingDeletes == 0 && !pagesRemaining) {
+      return true;
+    }
+
+    if (failures.length > 0) {
+      logger.debug("Found " + failures.length + " failed deletes, failing.");
+      return true;
     }
 
-    return self._deleteDocuments(docs).then(function(numDeleted) {
-      // Tick the progress bar
-      FirestoreDelete.progressBar.tick(numDeleted);
+    if (queue.length <= maxQueueSize && pagesRemaining && !pageIncoming) {
+      pageIncoming = true;
 
-      // Recurse to delete another batch
-      return self._recursiveBatchDelete();
-    });
+      self
+        ._getDescendantBatch(self.allDescendants, readBatchSize, lastDocName)
+        .then(function(docs) {
+          pageIncoming = false;
+
+          if (docs.length == 0) {
+            pagesRemaining = false;
+            return;
+          }
+
+          docs.forEach(function(doc) {
```

Comment:
This is https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/concat